binary method in defer

The following program `sample.go` triggers a panic:

```go
package main

import "sync"

type T struct {
    mu   sync.RWMutex
    name string
}   

func (t *T) get() string {
    t.mu.RLock()
    defer t.mu.RUnlock()
    return t.name
}   

var d = T{name: "test"}

func main() {
    println(d.get())
}
```

Expected result:
```console
$ go run ./sample.go
test
```

Got:
```console
$ yaegi ./sample.go
./sample.go:11:2: panic
./sample.go:19:10: panic
panic: reflect: Method index out of range [recovered]
        panic: reflect: Method index out of range [recovered]
        panic: reflect: Method index out of range

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc0002740a0, 0xc00028a300)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:114 +0x218
panic(0xc9eaa0, 0xf97b50)
        /usr/lib/go/src/runtime/panic.go:679 +0x1b2
github.com/containous/yaegi/interp.runCfg.func1(0xc0002741e0, 0xc000288200)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:114 +0x218
panic(0xc9eaa0, 0xf97b50)
        /usr/lib/go/src/runtime/panic.go:679 +0x1b2
reflect.Value.Method(0xd75bc0, 0xc00016e180, 0x199, 0x3, 0x20, 0xdd0720, 0xdd0701)
        /usr/lib/go/src/reflect/value.go:1312 +0x248
github.com/containous/yaegi/interp._defer.func1(0xc0002741e0, 0x1, 0x1, 0xc00017cc80)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:611 +0x64
github.com/containous/yaegi/interp._defer.func2(0xc0002741e0, 0xc0002824b0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:620 +0x8f
github.com/containous/yaegi/interp.runCfg(0xc000288200, 0xc0002741e0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:120 +0x7a
github.com/containous/yaegi/interp.call.func4(0xc0002740a0, 0xc000273570)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:808 +0xad7
github.com/containous/yaegi/interp.runCfg(0xc00028a300, 0xc0002740a0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:120 +0x7a
github.com/containous/yaegi/interp.(*Interpreter).run(0xc00026a580, 0xc000289900, 0xc000274000)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:98 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc00026a580, 0xc00027a000, 0xdc, 0x2dc, 0xc00027a000, 0xdc, 0x0, 0x1)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:367 +0x463
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:82 +0x42e
```

--- a/_test/defer4.go
+++ b/_test/defer4.go
@@ -1,0 +1,23 @@
+package main
+
+import "sync"
+
+type T struct {
+	mu   sync.RWMutex
+	name string
+}
+
+func (t *T) get() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.name
+}
+
+var d = T{name: "test"}
+
+func main() {
+	println(d.get())
+}
+
+// Output:
+// test

--- a/interp/ast.go
+++ b/interp/ast.go
@@ -200,7 +200,6 @@ const (
 	aCase
 	aCompositeLit
 	aDec
-	aDefer
 	aEqual
 	aGreater
 	aGreaterEqual
@@ -258,7 +257,6 @@ var actions = [...]string{
 	aCase:         "case",
 	aCompositeLit: "compositeLit",
 	aDec:          "--",
-	aDefer:        "defer",
 	aEqual:        "==",
 	aGreater:      ">",
 	aGetFunc:      "getFunc",
@@ -569,7 +567,7 @@ func (interp *Interpreter) ast(src, name string) (string, *node, error) {
 			st.push(addChild(&root, anc, pos, declStmt, aNop), nod)
 
 		case *ast.DeferStmt:
-			st.push(addChild(&root, anc, pos, deferStmt, aDefer), nod)
+			st.push(addChild(&root, anc, pos, deferStmt, aNop), nod)
 
 		case *ast.Ellipsis:
 			st.push(addChild(&root, anc, pos, ellipsisExpr, aNop), nod)

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1132,7 +1132,6 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					}
 				case n.typ.rtype.Kind() == reflect.Struct:
 					if field, ok := n.typ.rtype.FieldByName(n.child[1].ident); ok {
-						log.Println(n.index, "selector bin method #2.1")
 						n.typ = &itype{cat: valueT, rtype: field.Type}
 						n.val = field.Index
 						n.gen = getIndexSeq
@@ -1140,7 +1139,6 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 						// method lookup failed on type, now lookup on pointer to type
 						pt := reflect.PtrTo(n.typ.rtype)
 						if m2, ok2 := pt.MethodByName(n.child[1].ident); ok2 {
-							log.Println(n.index, "selector bin method #2.2")
 							n.val = m2.Index
 							n.gen = getIndexBinPtrMethod
 							n.typ = &itype{cat: valueT, rtype: m2.Type}

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -936,7 +936,7 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 			sc = sc.pop()
 			err = genRun(n)
 
-		case goStmt:
+		case deferStmt, goStmt:
 			wireChild(n)
 
 		case identExpr:
@@ -1132,6 +1132,7 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					}
 				case n.typ.rtype.Kind() == reflect.Struct:
 					if field, ok := n.typ.rtype.FieldByName(n.child[1].ident); ok {
+						log.Println(n.index, "selector bin method #2.1")
 						n.typ = &itype{cat: valueT, rtype: field.Type}
 						n.val = field.Index
 						n.gen = getIndexSeq
@@ -1139,6 +1140,7 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 						// method lookup failed on type, now lookup on pointer to type
 						pt := reflect.PtrTo(n.typ.rtype)
 						if m2, ok2 := pt.MethodByName(n.child[1].ident); ok2 {
+							log.Println(n.index, "selector bin method #2.2")
 							n.val = m2.Index
 							n.gen = getIndexBinPtrMethod
 							n.typ = &itype{cat: valueT, rtype: m2.Type}

--- a/interp/run.go
+++ b/interp/run.go
@@ -668,8 +668,7 @@ func call(n *node) {
 	}
 
 	if n.anc.kind == deferStmt {
-		// Resolve function and input args, but instead of executing,
-		// store function call in frame for deferred execution.
+		// Store function call in frame for deferred execution.
 		value = genFunctionWrapper(n.child[0])
 		n.exec = func(f *frame) bltn {
 			val := make([]reflect.Value, len(values)+1)
@@ -868,8 +867,7 @@ func callBin(n *node) {
 
 	switch {
 	case n.anc.kind == deferStmt:
-		// Resolve function and input args, but instead of executing,
-		// store function call in frame for deferred execution.
+		// Store function call in frame for deferred execution.
 		n.exec = func(f *frame) bltn {
 			val := make([]reflect.Value, l+1)
 			val[0] = value(f)

--- a/interp/run.go
+++ b/interp/run.go
@@ -29,7 +29,6 @@ var builtin = [...]bltnGenerator{
 	aCase:         _case,
 	aCompositeLit: arrayLit,
 	aDec:          dec,
-	aDefer:        _defer,
 	aEqual:        equal,
 	aGetFunc:      getFunc,
 	aGreater:      greater,
@@ -597,47 +596,6 @@ func genInterfaceWrapper(n *node, typ reflect.Type) func(*frame) reflect.Value {
 	}
 }
 
-func _defer(n *node) {
-	tnext := getExec(n.tnext)
-	values := make([]func(*frame) reflect.Value, len(n.child[0].child))
-	var method func(*frame) reflect.Value
-
-	for i, c := range n.child[0].child {
-		if c.typ.cat == funcT {
-			values[i] = genFunctionWrapper(c)
-		} else {
-			if c.recv != nil {
-				// defer a method on a binary obj
-				mi := c.val.(int)
-				m := genValue(c.child[0])
-				method = func(f *frame) reflect.Value { return m(f).Method(mi) }
-			}
-			values[i] = genValue(c)
-		}
-	}
-
-	if method != nil {
-		n.exec = func(f *frame) bltn {
-			val := make([]reflect.Value, len(values))
-			val[0] = method(f)
-			for i, v := range values[1:] {
-				val[i+1] = v(f)
-			}
-			f.deferred = append([][]reflect.Value{val}, f.deferred...)
-			return tnext
-		}
-	} else {
-		n.exec = func(f *frame) bltn {
-			val := make([]reflect.Value, len(values))
-			for i, v := range values {
-				val[i] = v(f)
-			}
-			f.deferred = append([][]reflect.Value{val}, f.deferred...)
-			return tnext
-		}
-	}
-}
-
 func call(n *node) {
 	goroutine := n.anc.kind == goStmt
 	var method bool
@@ -707,6 +665,22 @@ func call(n *node) {
 			j := n.findex + i
 			rvalues[i] = func(f *frame) reflect.Value { return f.data[j] }
 		}
+	}
+
+	if n.anc.kind == deferStmt {
+		// Resolve function and input args, but instead of executing,
+		// store function call in frame for deferred execution.
+		value = genFunctionWrapper(n.child[0])
+		n.exec = func(f *frame) bltn {
+			val := make([]reflect.Value, len(values)+1)
+			val[0] = value(f)
+			for i, v := range values {
+				val[i+1] = v(f)
+			}
+			f.deferred = append([][]reflect.Value{val}, f.deferred...)
+			return tnext
+		}
+		return
 	}
 
 	n.exec = func(f *frame) bltn {
@@ -893,8 +867,20 @@ func callBin(n *node) {
 	l := len(values)
 
 	switch {
+	case n.anc.kind == deferStmt:
+		// Resolve function and input args, but instead of executing,
+		// store function call in frame for deferred execution.
+		n.exec = func(f *frame) bltn {
+			val := make([]reflect.Value, l+1)
+			val[0] = value(f)
+			for i, v := range values {
+				val[i+1] = v(f)
+			}
+			f.deferred = append([][]reflect.Value{val}, f.deferred...)
+			return tnext
+		}
 	case n.anc.kind == goStmt:
-		// Execute function in a goroutine, discard results
+		// Execute function in a goroutine, discard results.
 		n.exec = func(f *frame) bltn {
 			in := make([]reflect.Value, l)
 			for i, v := range values {
@@ -904,7 +890,7 @@ func callBin(n *node) {
 			return tnext
 		}
 	case fnext != nil:
-		// Handle branching according to boolean result
+		// Handle branching according to boolean result.
 		n.exec = func(f *frame) bltn {
 			in := make([]reflect.Value, l)
 			for i, v := range values {

